src/librustdoc/html/markdown.rs

```diff
@@ -945,55 +938,36 @@ crate fn rust_code_blocks(md: &str) -> Vec<RustCodeBlock> {
 
                 if lang_string.rust {
                     in_rust_code_block = true;
-
-                    code_start = offset;
-                    code_block_start = match md[previous_offset..offset].find("```") {
-                        Some(fence_idx) => {
-                            is_fenced = true;
-                            previous_offset + fence_idx
-                        }
-                        None => {
-                            is_fenced = false;
-                            offset
-                        }
+                    code_block = Some(offset_range.clone());
+
+                    code_start = if !md[previous_offset.end..offset_range.start].ends_with("    ") {
```

Comment:
I don't think it is necessary to scan through the text to find the code block start/end now that it is using `into_offset_iter`.  It has the exact ranges provided by pulldown-cmark.  Perhaps something like this will be a little more reliable?

```diff
diff --git a/src/librustdoc/html/markdown.rs b/src/librustdoc/html/markdown.rs
index d10b131af8e..e1fc7075372 100644
--- a/src/librustdoc/html/markdown.rs
+++ b/src/librustdoc/html/markdown.rs
@@ -922,11 +922,11 @@ crate fn rust_code_blocks(md: &str) -> Vec<RustCodeBlock> {

     let mut p = Parser::new_ext(md, opts()).into_offset_iter();

-    let mut code_block = None;
+    let mut code_block_range = 0..0;
     let mut code_start = 0;
+    let mut code_end = 0;
     let mut is_fenced = false;
     let mut in_rust_code_block = false;
-    let mut previous_offset = Range { start: 0, end: 0 };
     while let Some((event, offset_range)) = p.next() {
         match event {
             Event::Start(Tag::CodeBlock(syntax)) => {
@@ -938,36 +938,23 @@ crate fn rust_code_blocks(md: &str) -> Vec<RustCodeBlock> {

                 if lang_string.rust {
                     in_rust_code_block = true;
-                    code_block = Some(offset_range.clone());
-
-                    code_start = if !md[previous_offset.end..offset_range.start].ends_with("    ") {
-                        is_fenced = true;
-                        offset_range.start + md[offset_range.clone()]
-                            .lines()
-                            .next()
-                            .map_or(0, |x| x.len() + 1)
-                    } else {
-                        is_fenced = false;
-                        offset_range.start
-                    };
+                    code_block_range = offset_range.clone();
+                    is_fenced = md[offset_range].starts_with("```");
+                }
+            }
+            Event::Text(_) if in_rust_code_block => {
+                // Once inside a code block, the Text event gives the contents
+                // of the code block.
+                if code_start == 0 {
+                    code_start = offset_range.start;
                 }
+                code_end = offset_range.end;
             }
             Event::End(Tag::CodeBlock(syntax)) if in_rust_code_block => {
                 in_rust_code_block = false;
-
-                let code_end = if is_fenced {
-                    let last_len = md[offset_range.clone()]
-                        .lines()
-                        .last()
-                        .filter(|l| l.ends_with("```"))
-                        .map_or(0, |l| l.len());
-                    offset_range.end - last_len
-                } else {
-                    offset_range.end
-                };
                 code_blocks.push(RustCodeBlock {
                     is_fenced,
-                    range: code_block.clone().unwrap(),
+                    range: code_block_range.clone(),
                     code: Range {
                         start: code_start,
                         end: code_end,
@@ -978,10 +965,10 @@ crate fn rust_code_blocks(md: &str) -> Vec<RustCodeBlock> {
                         None
                     },
                 });
+                code_start = 0;
             }
             _ => (),
         }
-        previous_offset = offset_range;
     }

     code_blocks
```

Reply:
Unfortunately it's not. I had this issue as well. :)

In the case the block code is indented and not started with backticks, then if we have backticks, it's text. We even have a test for this use case. Therefore, we need to check if the code block is preceded by whitespaces.

Reply:
Just to be clear, are you referring to [this test](https://github.com/rust-lang/rust/blob/d8014582b843d8722795b06586ef9c2042094963/src/test/rustdoc-ui/invalid-syntax.rs#L58-L62)?  If it is important to avoid adding the suggestion to add `text`, a modification of the patch above should work:  Add this to `Event::Text`:

`is_fenced = code_start != code_block_range.start;`

and remove `is_fenced` from `Tag::CodeBlock`.  That is, when the `Event::Text` has the same offset as `Event::Start(Tag::CodeBlock))`, then it should be an indented codeblock.

Reply:
That's a hack and not a good one unfortunately. Like I said below, the easiest way is to simply know beforehand the kind of code block this is. I sent a PR for it in pulldown-cmark. That'll allow us to just remove all of this.